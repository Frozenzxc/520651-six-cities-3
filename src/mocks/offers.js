import {getRandomArrayItem, getRandomIntegerNumber} from "../common";

const MAX_BEDROOMS = 6;
const MIN_BEDROOMS = 1;
const MAX_GUESTS = 10;
const MAX_PHOTOS = 6;
const MAX_RATING = 5;
const MIN_RATING = 1;

const CardTypes = [`apartment`, `room`, `house`, `hotel`];
const Options = [`Wifi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`];
const Photos = [
  `img/apartment-01.jpg`,
  `img/apartment-02.jpg`,
  `img/apartment-03.jpg`,
  `img/apartment-01.jpg`,
  `img/apartment-02.jpg`,
  `img/apartment-03.jpg`,
];

const generatePhotos = (photos) => {
  return photos
        .filter(() => Math.random() > 0.5)
        .slice(0, getRandomIntegerNumber(1, MAX_PHOTOS));
};

const generateOptions = (options) => {
  return options
        .filter(() => Math.random() > 0.5)
        .slice(0, getRandomIntegerNumber(0, Options.length));
};

export default [
  {
    bedroomsCount: 4,
    coordinates: [52.3909553943508, 4.85309666406198],
    description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras aliquet varius magna, non porta ligula feugiat eget. Fusce tristique felis at fermentum pharetra. Aliquam id orci ut lectus varius viverra.`,
    host: {
      name: `John`,
      avatar: `img/avatar-max.jpg`,
      super: false,
    },
    id: 1,
    maxGuests: 6,
    options: [`Wifi`, `Washing machine`, `Towels`, `Heating`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
    premium: false,
    price: 200,
    rating: 3,
    src: [`img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`],
    title: `Beautiful & luxurious apartment at great location`,
    type: `house`,
  },
  {
    bedroomsCount: 2,
    coordinates: [52.369553943508, 4.85309666406198],
    description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras aliquet varius magna, non porta ligula feugiat eget. Fusce tristique felis at fermentum pharetra. Aliquam id orci ut lectus varius viverra.`,
    host: {
      name: `Max`,
      avatar: `img/avatar-angelina.jpg`,
      super: false,
    },
    id: 2,
    maxGuests: 4,
    options: [`Wifi`, `Towels`, `Heating`, `Cabel TV`, `Fridge`],
    premium: true,
    price: 300,
    rating: 1,
    src: [`img/apartment-02.jpg`, `img/apartment-03.jpg`],
    title: `Wood and stone place`,
    type: `hotel`,
  },
  {
    bedroomsCount: 1,
    coordinates: [52.3909553943508, 4.929309666406198],
    description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras aliquet varius magna, non porta ligula feugiat eget. Fusce tristique felis at fermentum pharetra. Aliquam id orci ut lectus varius viverra.`,
    host: {
      name: `Bob`,
      avatar: `img/avatar-max.jpg`,
      super: true,
    },
    id: 3,
    maxGuests: 2,
    options: [`Wifi`, `Towels`, `Cabel TV`],
    premium: false,
    price: 450,
    rating: 3,
    src: [`img/apartment-03.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`],
    title: `Canal View Prinsengracht`,
    type: `room`,
  },
  {
    bedroomsCount: 2,
    coordinates: [52.3809553943508, 4.939309666406198],
    description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras aliquet varius magna, non porta ligula feugiat eget. Fusce tristique felis at fermentum pharetra. Aliquam id orci ut lectus varius viverra.`,
    host: {
      name: `Angelina`,
      avatar: `img/avatar-angelina.jpg`,
      super: false,
    },
    id: 4,
    maxGuests: 5,
    options: [`Wifi`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Cabel TV`],
    premium: false,
    price: 150,
    rating: 5,
    src: [`img/apartment-03.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`],
    title: `Nice, cozy, warm big bed apartment`,
    type: `apartment`,
  },
];
