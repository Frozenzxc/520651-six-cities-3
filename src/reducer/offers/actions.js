const ActionType = {
  BLOCK_FORM: `BLOCK_FORM`,
  LOAD_NEARBY_OFFERS: `LOAD_NEARBY_OFFERS`,
  LOAD_OFFERS: `LOAD_OFFERS`,
  LOAD_REVIEWS: `LOAD_REVIEWS`,
  SELECT_CARD: `SELECT_CARD`,
  SELECT_CITY: `SELECT_CITY`,
  SELECT_OFFER: `SELECT_OFFER`,
  SUCCESSFUL_POST_REVIEW: `SUCCESSFUL_POST_REVIEW`,
};

const ActionCreator = {

  blockForm: (isBlocked) => {
    return {
      type: ActionType.BLOCK_FORM,
      payload: isBlocked,
    };
  },

  loadNearbyOffers: (offers) => {
    return {
      type: ActionType.LOAD_NEARBY_OFFERS,
      payload: offers,
    };
  },

  loadOffers: (offers) => {
    return {
      type: ActionType.LOAD_OFFERS,
      payload: offers,
    };
  },

  loadReviews: (reviews) => {
    return {
      type: ActionType.LOAD_REVIEWS,
      payload: reviews,
    };
  },

  selectCard: (offer) => ({
    type: ActionType.SELECT_CARD,
    payload: offer.id,
  }),

  selectCity: (city) => ({
    type: ActionType.SELECT_CITY,
    payload: city,
  }),

  selectOffer: (offer) => ({
    type: ActionType.SELECT_OFFER,
    payload: offer,
  }),

  successfulPostReview: (isSuccess) => ({
    type: ActionType.SUCCESSFUL_POST_REVIEW,
    payload: isSuccess,
  })

};

export {ActionType, ActionCreator};
