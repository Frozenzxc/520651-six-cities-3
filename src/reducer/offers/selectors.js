import {createSelector} from "reselect";
import NameSpace from "../name-space";
import {MAX_NEARBY_OFFERS} from "../../const";

const getCurrentCity = (state) => {
  return state[NameSpace.OFFERS].currentCity;
};

const getActiveOffer = (state) => {
  return state[NameSpace.OFFERS].activeOffer;
};

const getFavoriteOffers = (state) => {
  return state[NameSpace.OFFERS].favoriteOffers;
};

const getNearbyOffers = (state) => {
  return state[NameSpace.OFFERS].nearbyOffers.slice(0, MAX_NEARBY_OFFERS);
};

const getOffers = (state) => {
  return state[NameSpace.OFFERS].offers;
};

const getReviews = (state) => {
  return state[NameSpace.OFFERS].reviews;
};

const getAvailableOffers = createSelector(
    getOffers,
    getCurrentCity,
    (offers, currentCity) => {
      return offers.filter((offer) => offer.city.name === currentCity);
    }
);

export {getCurrentCity, getActiveOffer, getNearbyOffers, getOffers, getReviews, getAvailableOffers, getFavoriteOffers};
