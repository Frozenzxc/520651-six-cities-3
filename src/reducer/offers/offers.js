import {extend} from "../../utils";
import {parseOffer, parseReview} from "../../utils";
import {ActionCreator, ActionType} from "../offers/actions";
import {ReviewPostingStatus} from "../../const";
import produce from "immer";

const getAvailableOffers = ((allOffers, currentCity) => allOffers.filter((offer) => offer.city.name === currentCity));

const initialState = {
  offers: [],
  availableOffers: [],
  activeOffer: null,
  currentCity: `Amsterdam`,
  favoriteOffers: [],
  isFavoritesLoading: true,
  isFormBlocked: false,
  isLoading: true,
  isNearbyOffersLoading: true,
  isReviewsLoading: true,
  nearbyOffers: [],
  reviews: [],
  reviewPostingStatus: null,
};

const Operation = {
  addToFavorite: (offer) => (dispatch, getState, api) => {
    return api.post(`/favorite/${offer.id}/${+!offer.isFavorite}`, {})
            .then((response) => {
              dispatch(ActionCreator.addToFavorite(response.data));
            });
  },


  loadFavoriteOffers: () => (dispatch, getState, api) => {
    return api.get(`/favorite`)
            .then((response) => {
              dispatch(ActionCreator.loadFavoriteOffers(response.data));
            });
  },

  loadNearbyOffers: (id) => (dispatch, getState, api) => {
    return api.get(`/hotels/${id}/nearby`)
            .then((response) => {
              dispatch(ActionCreator.loadNearbyOffers(response.data));
            });
  },

  loadOffers: () => (dispatch, getState, api) => {
    return api.get(`/hotels`)
            .then((response) => {
              dispatch(ActionCreator.loadOffers(response.data));
            });
  },

  loadReviews: (id) => (dispatch, getState, api) => {
    return api.get(`/comments/${id}`)
            .then((response) => {
              dispatch(ActionCreator.loadReviews(response.data));
            });
  },

  postReview: (id, formData) => (dispatch, getState, api) => {
    return api.post(`/comments/${id}`, {
      comment: formData.comment,
      rating: formData.rating,
    })
          .then((response) => {
            dispatch(ActionCreator.successfulPostReview(ReviewPostingStatus.POSTED));
            dispatch(ActionCreator.loadReviews(response.data));
            dispatch(ActionCreator.blockForm(false));
          })
        .catch((err) => {
          dispatch(ActionCreator.successfulPostReview(ReviewPostingStatus.ERROR));
          dispatch(ActionCreator.blockForm(false));
          throw err;
        });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.ADD_TO_FAVORITE:
      const parsedOffer = parseOffer(action.payload);
      const index = state.availableOffers.findIndex((offer) => offer.id === parsedOffer.id);
      const nextState = produce(state, (draft) => {
        draft.availableOffers[index].isFavorite = parsedOffer.isFavorite;
        if (!parsedOffer.isFavorite) {
          const offerIndex = state.favoriteOffers.findIndex((offer) => offer.id === parsedOffer.id);
          draft.favoriteOffers.splice([offerIndex], 1);
        }
      });
      return extend(state, {
        availableOffers: nextState.availableOffers,
        favoriteOffers: nextState.favoriteOffers,
      });

    case ActionType.BLOCK_FORM:
      return extend(state, {
        isFormBlocked: action.payload,
      });

    case ActionType.LOAD_FAVORITE_OFFERS:
      let parsedFavoriteOffers = action.payload.map((offer) => parseOffer(offer));
      return extend(state, {
        favoriteOffers: parsedFavoriteOffers,
        isFavoritesLoading: false,
      });

    case ActionType.LOAD_NEARBY_OFFERS:
      let parsedNearbyOffers = action.payload.map((offer) => parseOffer(offer));
      return extend(state, {
        nearbyOffers: parsedNearbyOffers,
        isNearbyOffersLoading: false,
      });

    case ActionType.LOAD_OFFERS:
      let parsedOffers = action.payload.map((offer) => parseOffer(offer));
      return extend(state, {
        availableOffers: getAvailableOffers(parsedOffers, parsedOffers[0].city.name),
        currentCity: parsedOffers[0].city.name,
        isLoading: false,
        offers: parsedOffers,
      });

    case ActionType.LOAD_REVIEWS:
      let parsedReviews = action.payload.map((review) => parseReview(review));
      return extend(state, {
        isReviewsLoading: false,
        reviews: parsedReviews,
      });

    case ActionType.SELECT_CITY:
      return extend(state, {
        currentCity: action.payload,
        availableOffers: getAvailableOffers(state.offers, action.payload),
      });

    case ActionType.SELECT_OFFER:
      return extend(state, {
        activeOffer: action.payload,
      });

    case ActionType.SUCCESSFUL_POST_REVIEW:
      return extend(state, {
        reviewPostingStatus: action.payload,
      });
  }

  return state;
};

export {reducer, Operation, ActionType, ActionCreator};
