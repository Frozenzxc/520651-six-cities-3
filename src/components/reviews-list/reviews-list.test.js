import React from 'react';
import renderer from 'react-test-renderer';
import offers from "../../test-mocks/reviews";
import ReviewsList from "./reviews-list";

const reviews = offers[0].reviews;

it(`ReviewsList is rendered correctly`, () => {
  const tree = renderer.create(
      <ReviewsList
        reviews={reviews}
      />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
