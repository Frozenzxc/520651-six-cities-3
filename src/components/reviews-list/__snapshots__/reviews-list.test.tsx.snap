// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsList is rendered correctly 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/8.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Emely
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!
        </p>
        <time
          className="reviews__time"
          dateTime="2020-02-23"
        >
          February 23, 2020
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/5.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Corey
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          The house is very good, very happy, hygienic and simple living conditions around it are also very good. I hope to have the opportunity to come back. Thank you.
        </p>
        <time
          className="reviews__time"
          dateTime="2020-02-23"
        >
          February 23, 2020
        </time>
      </div>
    </li>
  </ul>
</section>
`;
